File: 20242Q/poo/laboratorio/tp05/_soluciones/ej3/figures/Rectangle.java
package ar.edu.itba.poo.soltp5.ej3.figures;

public class Rectangle<P extends Point> extends Figure {

    protected final P topLeft, bottomRight;

    public Rectangle(P topLeft, P bottomRight) {
        this.topLeft = topLeft;
        this.bottomRight = bottomRight;
    }

    public double base() {
        return Math.abs(topLeft.getX() - bottomRight.getX());
    }

    public double height() {
        return Math.abs(topLeft.getY() - bottomRight.getY());
    }

    @Override
    public double area() {
        return base() * height();
    }

    @Override
    public double perimeter() {
        return (base() + height()) * 2;
    }

    @Override
    public String toString() {
        return "Rectángulo [ %s , %s ]".formatted(topLeft, bottomRight);
    }

}

File: 20242Q/poo/laboratorio/tp08/ej4/rectangle.rb
require_relative 'figure'

class Rectangle < Figure
  def initialize(top_left, bottom_right)
    @top_left = top_left
    @bottom_right = bottom_right
  end

  def area
    base * height
  end

  def perimeter
    (base + height) * 2
  end

  def to_s
    "Rectángulo [ #{@top_left} , #{@bottom_right} ]"
  end

  private

  def base
    (@top_left.x - @bottom_right.x).abs
  end

  def height
    (@top_left.y - @bottom_right.y).abs
  end
end
